File: gradle-plugins/gradletestapps/java-one-project/src/test/java/com/example/execution/javaoneproject/PluginExecutionSmokeTest.java
package com.example.execution.javaoneproject;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.jar.JarFile;
import org.junit.jupiter.api.Test;
import viaduct.service.BasicViaductFactory;
import viaduct.service.api.ExecutionInput;
import viaduct.service.api.ExecutionResult;
import viaduct.service.api.Viaduct;

/**
 * End-to-end coverage of the Java module plugin in a single-project application: schema assembly,
 * Java resolver-base codegen, the APT-driven registry pipeline, and execution against a real
 * Viaduct instance bootstrapped from the packaged registry.
 */
public class PluginExecutionSmokeTest {

  private static final String REGISTRY_ENTRY =
      "META-INF/viaduct/modules/com.example.execution.javaoneproject.resolvers.json";

  private final Path buildDir = Path.of(System.getProperty("projectBuildDir"));

  @Test
  void generatedSchemaAndResolverOutputsExist() throws IOException {
    assertExists(buildDir.resolve("viaduct/centralSchema/BUILTIN_SCHEMA.graphqls"));
    Path baseSchemaFile = buildDir.resolve("viaduct/centralSchema/schemabase/directives.graphqls");
    Path commonSchemaFile = buildDir.resolve("viaduct/centralSchema/common/common.graphqls");
    Path partitionSchemaFile =
        buildDir.resolve("viaduct/centralSchema/partition/resolvers/graphql/schema.graphqls");

    assertExists(baseSchemaFile);
    assertExists(commonSchemaFile);
    assertExists(partitionSchemaFile);

    Path resolverBases =
        buildDir.resolve(
            "generated-sources/viaduct/javaResolverBases/"
                + "com/example/execution/javaoneproject/resolvers/resolverbases");
    assertExists(resolverBases.resolve("QueryResolvers.java"));
    assertExists(resolverBases.resolve("MutationResolvers.java"));

    assertTrue(Files.readString(baseSchemaFile).contains("directive @javaOneProjectBase"));
    assertTrue(Files.readString(commonSchemaFile).contains("directive @javaOneProjectCommon"));

    String partitionSchema = Files.readString(partitionSchemaFile);
    assertTrue(partitionSchema.contains("greeting: String"));
    assertTrue(partitionSchema.contains("echo(message: String!): String"));
  }

  @Test
  void tenantModuleConfigContainsAptExtractedResolvers() throws IOException {
    Path configFile = buildDir.resolve("generated-resources/viaduct-registry/" + REGISTRY_ENTRY);

    assertExists(configFile);

    String contents = Files.readString(configFile);
    assertTrue(contents.contains("GreetingResolver"));
    assertTrue(contents.contains("AuthorResolver"));
    assertTrue(contents.contains("EchoMutationResolver"));
    assertTrue(contents.contains("viaduct.java.runtime.bootstrap.ViaductJavaExecutorFactory"));
  }

  /**
   * javac's {@code Filer} writes the per-source descriptors into the class output rather than to a
   * dedicated codegen directory, so the assembled registry and these build intermediates share a
   * tree. Only the former belongs in the jar.
   */
  @Test
  void jarShipsAssembledRegistryWithoutRawDescriptors() throws IOException {
    assertExists(buildDir.resolve("classes/java/main/viaduct-registry"));

    String moduleJar = System.getProperty("moduleJar");
    try (JarFile jar = new JarFile(moduleJar)) {
      assertNotNull(
          jar.getJarEntry(REGISTRY_ENTRY),
          "Expected the assembled registry at " + REGISTRY_ENTRY + " in " + moduleJar);
      assertFalse(
          jar.stream().anyMatch(entry -> entry.getName().startsWith("viaduct-registry/")),
          "Expected no raw APT descriptors in " + moduleJar);
    }
  }

  @Test
  void queriesAndMutationsExecuteThroughViaduct() {
    Viaduct viaduct = BasicViaductFactory.create();

    ExecutionResult queryResult =
        viaduct.executeAsync(ExecutionInput.create("query { greeting author }")).join();
    assertTrue(
        queryResult.getErrors().isEmpty(),
        "Expected query execution without errors: " + queryResult.getErrors());
    assertEquals(
        Map.of(
            "greeting", "hello from java-one-project",
            "author", "gradletestapps"),
        queryResult.getData());

    ExecutionResult mutationResult =
        viaduct
            .executeAsync(ExecutionInput.create("mutation { echo(message: \"plugin e2e\") }"))
            .join();
    assertTrue(
        mutationResult.getErrors().isEmpty(),
        "Expected mutation execution without errors: " + mutationResult.getErrors());
    assertEquals(Map.of("echo", "plugin e2e"), mutationResult.getData());
  }

  private void assertExists(Path path) {
    assertTrue(Files.exists(path), "Expected generated output to exist: " + path);
  }
}

File: gradle-plugins/gradletestapps/one-project/src/test/kotlin/com/example/execution/oneproject/PluginExecutionSmokeTest.kt
package com.example.execution.oneproject

import java.nio.file.Path
import kotlin.io.path.exists
import kotlin.io.path.readText
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import viaduct.service.BasicViaductFactory
import viaduct.service.api.ExecutionInput

class PluginExecutionSmokeTest {
    private val buildDir = Path.of(System.getProperty("projectBuildDir"))

    @Test
    fun generatedSchemaAndResolverOutputsExist() {
        assertExists(buildDir.resolve("viaduct/centralSchema/BUILTIN_SCHEMA.graphqls"))
        val baseSchemaFile = buildDir.resolve("viaduct/centralSchema/schemabase/directives.graphqls")
        val commonSchemaFile = buildDir.resolve("viaduct/centralSchema/common/common.graphqls")
        val partitionSchemaFile = buildDir.resolve("viaduct/centralSchema/partition/resolvers/graphql/schema.graphqls")

        assertExists(baseSchemaFile)
        assertExists(commonSchemaFile)
        assertExists(partitionSchemaFile)
        assertExists(
            buildDir.resolve(
                "generated-sources/viaduct/resolverBases/" +
                    "com/example/execution/oneproject/resolvers/QueryResolvers.kt"
            )
        )
        assertExists(
            buildDir.resolve(
                "generated-sources/viaduct/resolverBases/" +
                    "com/example/execution/oneproject/resolvers/MutationResolvers.kt"
            )
        )

        assertTrue(baseSchemaFile.readText().contains("directive @oneProjectBase"))
        assertTrue(commonSchemaFile.readText().contains("directive @oneProjectCommon"))
        assertTrue(partitionSchemaFile.readText().contains("greeting: String"))
        assertTrue(partitionSchemaFile.readText().contains("echo(message: String!): String"))
    }

    @Test
    fun tenantModuleConfigContainsKspExtractedResolvers() {
        val configFile = buildDir.resolve(
            "generated-resources/viaduct-registry/" +
                "META-INF/viaduct/modules/com.example.execution.oneproject.resolvers.json"
        )

        assertExists(configFile)

        val contents = configFile.readText()
        assertTrue(contents.contains("GreetingResolver"))
        assertTrue(contents.contains("AuthorResolver"))
        assertTrue(contents.contains("EchoMutationResolver"))
    }

    @Test
    fun queriesAndMutationsExecuteThroughViaduct() {
        val viaduct = BasicViaductFactory.create()

        val queryResult = viaduct.executeAsync(
            ExecutionInput.create("query { greeting author }")
        ).join()
        assertTrue(queryResult.errors.isEmpty(), "Expected query execution without errors: ${queryResult.errors}")
        assertEquals(
            mapOf(
                "greeting" to "hello from one-project",
                "author" to "gradletestapps",
            ),
            queryResult.getData(),
        )

        val mutationResult = viaduct.executeAsync(
            ExecutionInput.create("""mutation { echo(message: "plugin e2e") }""")
        ).join()
        assertTrue(mutationResult.errors.isEmpty(), "Expected mutation execution without errors: ${mutationResult.errors}")
        assertEquals(mapOf("echo" to "plugin e2e"), mutationResult.getData())
    }

    @Test
    fun invalidSyntaxProducesParseError() {
        val viaduct = BasicViaductFactory.create()

        val result = viaduct.executeAsync(
            ExecutionInput.create("query { }")
        ).join()

        assertNull(result.getData())
        assertTrue(result.errors.isNotEmpty(), "Expected parse errors for invalid syntax")
        assertTrue(result.errors.first().message.contains("Invalid syntax"))
    }

    private fun assertExists(path: Path) {
        assertTrue(path.exists(), "Expected generated output to exist: $path")
    }
}
